<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2019-12-12" />

<title>Experiments with Prior Parameter Estimation</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CAUSE</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Introduction</a>
</li>
<li>
  <a href="ldl_cad.html">Software Tutorial</a>
</li>
<li>
  <a href="simulations.html">Simulations</a>
</li>
<li>
  <a href="gwas_pairs_2.html">Pairs of GWAS Traits</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/cause">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Experiments with Prior Parameter Estimation</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2019-12-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-12-13
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>cause/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0.9000). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181014code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181014)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181014code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181014)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997causetree9a532644804701ea44eb507bb0fd1aa88316679ctargetblank9a53264a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/cause/tree/9a532644804701ea44eb507bb0fd1aa88316679c" target="_blank">9a53264</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997causetree9a532644804701ea44eb507bb0fd1aa88316679ctargetblank9a53264a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    sim_results/
    Ignored:    src/RcppExports.o
    Ignored:    src/cause.so
    Ignored:    src/log_likelihood_functions.o

Untracked files:
    Untracked:  analysis/figure/ldl_cad.Rmd/
    Untracked:  analysis/figure/little_test.Rmd/
    Untracked:  analysis/ldl_cad_future.Rmd
    Untracked:  cause.Rcheck/
    Untracked:  docs/cause.bib
    Untracked:  docs/figure/cause_figure_1_standalone.pdf
    Untracked:  gwas_data/
    Untracked:  ll_v7_notes.Rmd
    Untracked:  man/est_cause_params_future2.Rd
    Untracked:  temp.txt
    Untracked:  tests/
    Untracked:  upd_grid_test/
    Untracked:  upd_mixsqp_test/

Unstaged changes:
    Modified:   example_data/LDL_CAD_merged.RDS

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/9a532644804701ea44eb507bb0fd1aa88316679c/analysis/grid_estimation.Rmd" target="_blank">9a53264</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-12-13
</td>
<td>
wflow_publish(“analysis/grid_estimation.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>These experiments came about in the course of looking at the behavior of CAUSE for very high sample size. The question I set out with was, when <span class="math inline">\(q\)</span> is large (say 0.4) and <span class="math inline">\(\gamma = 0\)</span>, will CAUSE eventually be able to distinguish this from a causal scenario if the signal is strong enough? We would hope this would be the case. In the results we present in the paper, the highest sample size we use is 40,000 for each trait (this is the high power <span class="math inline">\(M\)</span>/high power <span class="math inline">\(Y\)</span> condition) and the false positive rate is pretty large for larger values of <span class="math inline">\(q\)</span>.</p>
<p>I started turning the sample size up to see what would happen. I looked at sizes of <span class="math inline">\(10^5\)</span>, <span class="math inline">\(10^6\)</span> and <span class="math inline">\(10^8\)</span>. At <span class="math inline">\(10^8\)</span> CAUSE was able to correctly estimate all the parameters and reject the causal model. However, at <span class="math inline">\(10^6\)</span> it was still getting false positives. When I visually inspected the data produced with the <span class="math inline">\(10^6\)</span> sample size, the signal looked clear to me and I thought that CAUSE should be doing a better job. This lead me to wonder if the prior distribution we estimate for SNP effects was too wide, making it hard to distinguish models.</p>
</div>
<div id="snp-effect-prior-estimation-procedurecause-model-quick-review" class="section level2">
<h2>SNP effect prior estimation procedure/CAUSE model quick review</h2>
<p>Briefly stated, the CAUSE model for summary statistics is</p>
<p><span class="math display">\[
\begin{pmatrix} \hat{\beta}_{M,j}\\ \hat{\beta}_{Y,j} \end{pmatrix} \sim N\left(\begin{pmatrix} \beta_{M,j}\ \\ \beta_{Y,j} \end{pmatrix}, 
\begin{pmatrix} s_{M,j}^2 &amp; \rho s_{M,j} s_{Y,j} \\ \rho s_{M,j}s_{Y,j} &amp; s_{Y,j}^2 \end{pmatrix}\right)
\]</span> <span class="math display">\[
\beta_{Y,j} = \gamma \beta_{M,j} + \eta Z_j \beta_{M,j} + \theta_j 
\]</span> <span class="math display">\[
Z_j \sim Binom(q) 
\]</span> <span class="math display">\[
\begin{pmatrix} \beta_{M,j}\\ \theta_j \end{pmatrix} \sim  \sum_{k=0}^{K} \pi_k N\left(\begin{pmatrix}0 \\0  \end{pmatrix}, 
\begin{pmatrix} t_{k,1}^2 &amp; 0 \\ 0 &amp; t_{k,2}^2 \end{pmatrix}\right)
\]</span> We assume the pairs <span class="math inline">\((t_{k,1}, t_{k,2})\)</span> are fixed in advanced (referred to as “the grid” or “variance pair candidates”). The CAUSE model has the following unknown parameters: <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\eta\)</span>, <span class="math inline">\(q\)</span>, <span class="math inline">\(\rho\)</span>, and <span class="math inline">\(\pi\)</span>. The first four are all scalars and <span class="math inline">\(\pi \in \mathbb{R}^{K+1}\)</span> and deterimines the distribution of SNP effects.</p>
<p>In our current method of estimation, we estimate <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\rho\)</span> first and fix these and then compute posterior distributions for <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\eta\)</span>, and <span class="math inline">\(q\)</span>. This estimation step is what I am investigating here. In what I will call “the original” estimation scheme, <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\rho\)</span> are obtained by maximizing the likelihood with <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\eta\)</span>, and <span class="math inline">\(q\)</span> all fixed at zero. One consequence of this is that, if <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\eta\)</span>, and <span class="math inline">\(q\)</span> are not all zero, more weight will be given to larger variances than is necessary. In this document I will discuss these alternatives:</p>
<ul>
<li>“fixed oracle”: <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\rho\)</span> are estimated by maximizing the likelihood with <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\eta\)</span>, and <span class="math inline">\(q\)</span> fixed at their true values. This is obviously only possible in simulations but is a useful comparison.</li>
<li>“fixed est”: <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\rho\)</span> are estimated by maximizing the likelihood with <span class="math inline">\(\gamma\)</span> fixed at the point estimate obtained from the weighted mode MR estimate, and <span class="math inline">\(\eta\)</span>, and <span class="math inline">\(q\)</span> fixed at zero.</li>
<li>“eqg”: In the first two rounds of coordinate descent, <span class="math inline">\(\eta\)</span>, <span class="math inline">\(q\)</span>, and <span class="math inline">\(\gamma\)</span> are also updated (in that order). After these iterations, they are fixed.</li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> <span class="math inline">\(\rightarrow\)</span> The investigation came about because I discovered that using a scheme similar to eqg, I could almost entirely avoid false positives with large <span class="math inline">\(q\)</span> at lower sample sizes. However (spoiler) this method also lead to much lower power and therefore wasn’t acceptable. To understand why, I then implemented the fixed oracle and fixed est methods. I found that these also had low power which surprised me, especially for fixed oracle which I would expect to do well.</p>
</div>
<div id="comparisons-with-fixed-oracle" class="section level2">
<h2>Comparisons with fixed oracle</h2>
<p>To understand the effect of the fixed oracle parameters I will go through a few examples.</p>
<div id="high-sample-size-q0.5-gamma0" class="section level3">
<h3>High sample size, <span class="math inline">\(q=0.5\)</span>, <span class="math inline">\(\gamma=0\)</span></h3>
<pre class="r"><code>res_orig &lt;- readRDS(&quot;upd_grid_test/high_sample_q0.5__res_orig.rds&quot;)
summary(res_orig)</code></pre>
<pre><code>p-value testing that causal model is a better fit:  0.0096 
Posterior medians and  95 % credible intervals:
     model     gamma               eta                 
[1,] &quot;Sharing&quot; NA                  &quot;0.15 (0.12, 0.17)&quot; 
[2,] &quot;Causal&quot;  &quot;0.11 (0.08, 0.13)&quot; &quot;0.02 (-0.15, 0.17)&quot;
     q                  
[1,] &quot;0.71 (0.59, 0.83)&quot;
[2,] &quot;0.06 (0, 0.33)&quot;   </code></pre>
<pre class="r"><code>res_fo &lt;- readRDS(&quot;upd_grid_test/high_sample_q0.5___res_fo.rds&quot;)
summary(res_fo)</code></pre>
<pre><code>p-value testing that causal model is a better fit:  0.69 
Posterior medians and  95 % credible intervals:
     model     gamma            eta                 q                  
[1,] &quot;Sharing&quot; NA               &quot;0.22 (0.21, 0.23)&quot; &quot;0.5 (0.46, 0.54)&quot; 
[2,] &quot;Causal&quot;  &quot;0.01 (0, 0.23)&quot; &quot;0.2 (-0.22, 0.22)&quot; &quot;0.49 (0.45, 0.54)&quot;</code></pre>
<p>Using the original parameters we are getting a false positive. In the causal model we are estimating a causal effect that is between the true confounder effect (0.22) and 0. The sharing model is also not quite accurate, the proportion <span class="math inline">\(q\)</span> is too large, it is also trying to fit this comporomise causal model. Using the “oracle” grid we get the correct estimates from both models and no false positive. We can compare the ELPD contributions.</p>
<p><img src="figure/grid_estimation.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="true-positive-gamma-neq-0-no-shared-factor" class="section level2">
<h2>True positive <span class="math inline">\(\gamma \neq 0\)</span>, no shared factor</h2>
<p>My intuition is that I would also expect this method to improve power, but that was not my finding. Here is an example with a sample size of 40,000 for each trait and a true causal effect.</p>
<pre class="r"><code>res_orig &lt;- readRDS(&quot;upd_grid_test/pwr_example__res_orig.rds&quot;)
summary(res_orig)</code></pre>
<pre><code>p-value testing that causal model is a better fit:  0.00011 
Posterior medians and  95 % credible intervals:
     model     gamma              eta                  
[1,] &quot;Sharing&quot; NA                 &quot;0.18 (0.13, 0.23)&quot;  
[2,] &quot;Causal&quot;  &quot;0.13 (0.1, 0.16)&quot; &quot;-0.02 (-0.66, 0.37)&quot;
     q                  
[1,] &quot;0.57 (0.38, 0.75)&quot;
[2,] &quot;0.03 (0, 0.25)&quot;   </code></pre>
<pre class="r"><code>res_fo &lt;- readRDS(&quot;upd_grid_test/pwr_example___res_fo.rds&quot;)
summary(res_fo)</code></pre>
<pre><code>p-value testing that causal model is a better fit:  0.19 
Posterior medians and  95 % credible intervals:
     model     gamma               eta                  
[1,] &quot;Sharing&quot; NA                  &quot;0.22 (0.18, 0.27)&quot;  
[2,] &quot;Causal&quot;  &quot;0.15 (0.11, 0.19)&quot; &quot;-0.12 (-0.79, 0.36)&quot;
     q                  
[1,] &quot;0.56 (0.42, 0.71)&quot;
[2,] &quot;0.05 (0, 0.27)&quot;   </code></pre>
<p>In this case the fits are almost the same. The sharing model estimates a relatively large proportion of shared variants and the causal model estimates a causal effect and almost no sharing. The estimates of <span class="math inline">\(\eta\)</span> are slightly different but the posterior in both cases is very wide. However, the significance is very different. Using the original parameters the <span class="math inline">\(p\)</span>-value is 0.00011 while using the “fixed oracle” parameters it is 0.19. Plots can show why.</p>
<p><img src="figure/grid_estimation.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Red points are votes for the causal model while blue points are votes against. The fixed oracle parameters aren’t wide enough to accomodate the points with the largest excursions from the point estimate. These points (blue upper right on the right hand plot) are fit better by the sharing model which has a larger effect for the shared factor.</p>
<p>The plot below shows the difference in elpd contributions. Red points are more in favor of the causal model using the original parameters while blue points are more strongly in favor of the causal model using the fixed oracle paramters.</p>
<p><img src="figure/grid_estimation.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="simulations" class="section level2">
<h2>Simulations</h2>
<p>As we might expect form the examples above, in simulations, fixed oracle parameters almost entirely eliminate false positives. However, the power is significantly reduced. What can we do to retain power but still eliminate false positives and get good parameter estimates? Note that the fixed estimate parameters (“CAUSE-fe”) are also included in these plots. The fixed oracle are “CAUSE-fo”. I haven’t discussed these but the perform fairly simiarly to the oracle estimates but don’t control the false positive rate quite as well. I got somewhat better false positive results from the eqg method which also has much worse power.</p>
<p><img src="pwr_2019-12-13.png" style="display: block; margin: auto;" /></p>
<p><img src="fp_2019-12-13.png" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] gridExtra_2.3    forcats_0.4.0    stringr_1.4.0    dplyr_0.8.3     
 [5] purrr_0.3.3      readr_1.3.1      tidyr_1.0.0      tibble_2.1.3    
 [9] ggplot2_3.2.1    tidyverse_1.2.1  cause_0.3.0.0235

loaded via a namespace (and not attached):
 [1] httr_1.4.1           jsonlite_1.6         foreach_1.4.7       
 [4] modelr_0.1.5         RcppParallel_4.4.4   assertthat_0.2.1    
 [7] mixsqp_0.1-97        cellranger_1.1.0     yaml_2.2.0          
[10] numDeriv_2016.8-1.1  pillar_1.4.2         backports_1.1.5     
[13] lattice_0.20-38      glue_1.3.1           digest_0.6.23       
[16] rvest_0.3.4          colorspace_1.4-1     htmltools_0.3.6     
[19] Matrix_1.2-17        pkgconfig_2.0.3      broom_0.5.2         
[22] haven_2.1.1          scales_1.1.0         intervals_0.15.1    
[25] whisker_0.4          git2r_0.26.1         generics_0.0.2      
[28] farver_2.0.1         withr_2.1.2          ashr_2.2-32         
[31] lazyeval_0.2.2       cli_2.0.0            magrittr_1.5        
[34] crayon_1.3.4         readxl_1.3.1         evaluate_0.14       
[37] fs_1.3.1             fansi_0.4.0          doParallel_1.0.15   
[40] nlme_3.1-141         MASS_7.3-51.4        xml2_1.2.2          
[43] truncnorm_1.0-8      tools_3.6.1          loo_2.1.0           
[46] hms_0.5.2            lifecycle_0.1.0      matrixStats_0.55.0  
[49] munsell_0.5.0        compiler_3.6.1       rlang_0.4.2         
[52] grid_3.6.1           iterators_1.0.12     rstudioapi_0.10     
[55] labeling_0.3         rmarkdown_1.15       gtable_0.3.0        
[58] codetools_0.2-16     R6_2.4.1             lubridate_1.7.4     
[61] knitr_1.24           zeallot_0.1.0        workflowr_1.4.0.9000
[64] rprojroot_1.3-2      stringi_1.4.3        pscl_1.5.2          
[67] parallel_3.6.1       SQUAREM_2017.10-1    Rcpp_1.0.3          
[70] vctrs_0.2.0          tidyselect_0.2.5     xfun_0.9            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
