<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2019-07-10" />

<title>Simulations</title>

<script src="site_libs/header-attrs-2.3/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CAUSE</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Introduction</a>
</li>
<li>
  <a href="ldl_cad.html">Software Tutorial</a>
</li>
<li>
  <a href="simulations.html">Simulations</a>
</li>
<li>
  <a href="gwas_pairs_2.html">Pairs of GWAS Traits</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/cause">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Simulations</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2019-07-10</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-07-13
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>cause/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181014code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181014)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181014code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181014)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997causetreeb785efc43969e2444651864f6919292bead187fbtargetblankb785efca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/cause/tree/b785efc43969e2444651864f6919292bead187fb" target="_blank">b785efc</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997causetreeb785efc43969e2444651864f6919292bead187fbtargetblankb785efca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jean997/cause/tree/b785efc43969e2444651864f6919292bead187fb" target="_blank">b785efc</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    sim_results/
    Ignored:    src/RcppExports.o
    Ignored:    src/cause.so
    Ignored:    src/log_likelihood_functions.o

Untracked files:
    Untracked:  analysis/mrcieu.Rmd
    Untracked:  cause.Rcheck/
    Untracked:  gwas_data/
    Untracked:  ll_v7_notes.Rmd
    Untracked:  ll_v7_notes.html
    Untracked:  tests/

Unstaged changes:
    Deleted:    analysis/figure/simulations.Rmd/plot_fp-1.png
    Deleted:    analysis/figure/simulations.Rmd/plot_power-1.png
    Deleted:    analysis/figure/simulations.Rmd/roc_curves-1.png

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/simulations.Rmd</code>) and HTML (<code>docs/simulations.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/b785efc43969e2444651864f6919292bead187fb/analysis/simulations.Rmd" target="_blank">b785efc</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-07-13
</td>
<td>
wflow_publish(“analysis/simulations.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/42dca679cef0e467f3aaf8030e05bddc2365914f/docs/simulations.html" target="_blank">42dca67</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-12-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/104aa4390f620c7abe99a674c194a7a75f87c164/docs/simulations.html" target="_blank">104aa43</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-10-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/d5d0a475811aca2ad52b5141ef0a4c896037b98a/analysis/simulations.Rmd" target="_blank">d5d0a47</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-10-15
</td>
<td>
wflow_publish(c(“analysis/simulations.Rmd”))
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/12413f743c9642f36d2cf5a34a14105725a214e1/analysis/simulations.Rmd" target="_blank">12413f7</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-25
</td>
<td>
small typo
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/9de7de67ae150f8230419f2137ec7fe0822dc2a3/docs/simulations.html" target="_blank">9de7de6</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/112bcdee0633a4f6d155e402104e1527eca2d19a/analysis/simulations.Rmd" target="_blank">112bcde</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-24
</td>
<td>
smally typo
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/b1154146bef8ee5f0a4b1471490734da51ba2f29/docs/simulations.html" target="_blank">b115414</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/6aecf9160dae8320d1281d9d5ae5bf0d1d7e1c78/analysis/simulations.Rmd" target="_blank">6aecf91</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-24
</td>
<td>
change some wording and add extraction code
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/8bbb5162069dbefc21f0fca851275dcfed218a96/docs/simulations.html" target="_blank">8bbb516</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/86ffddd20efe2d2834dc5b682ada896f97c55bdd/analysis/simulations.Rmd" target="_blank">86ffddd</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-15
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/simulations.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/1074c6128db95e1abb96fa5697c39ef469e1d266/docs/simulations.html" target="_blank">1074c61</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/7e3c495fd74213919e52a13d592b057ab27bf30b/analysis/simulations.Rmd" target="_blank">7e3c495</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-10
</td>
<td>
more plotting details
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/aeeacea9a4e3e1032e79a0d1271787aa47c2015d/analysis/simulations.Rmd" target="_blank">aeeacea</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-10
</td>
<td>
lots of plotting details. still working
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/d2f0f741720daa4b2b05f7c7da4d55cdb0cd4a0b/docs/simulations.html" target="_blank">d2f0f74</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/6c171b2cc9a8d040138d79e1b404a65de9f2b75d/analysis/simulations.Rmd" target="_blank">6c171b2</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-09
</td>
<td>
wflow_publish(“analysis/simulations.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/165f2c81412fee8c44f95eb925b25891fec0d55e/docs/simulations.html" target="_blank">165f2c8</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-09
</td>
<td>
Revert “Build site.”
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/2666541e7075abd688e073d0df535398514855ed/docs/simulations.html" target="_blank">2666541</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/16285d63a5d60c078293c14b16b8ed6df1d09c5a/analysis/simulations.Rmd" target="_blank">16285d6</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-08
</td>
<td>
wflow_publish(“analysis/simulations.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/a1b74c5840b59877e79510e747131457961c9029/docs/simulations.html" target="_blank">a1b74c5</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/31a7da8399e15fed79099247f949411141db4e04/analysis/simulations.Rmd" target="_blank">31a7da8</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-08
</td>
<td>
wflow_publish(“analysis/simulations.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/0dd65ccae04d204000f8e640b175ef011b015839/docs/simulations.html" target="_blank">0dd65cc</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-06-25
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/b78e77f8d72199464465a98e0b3cfe2a6f9503c5/analysis/simulations.Rmd" target="_blank">b78e77f</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-06-25
</td>
<td>
wflow_publish(files = c(“analysis/simulations.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/cause/286f4e9e0a432565b96c38bba3a1b155c02f9a78/docs/simulations.html" target="_blank">286f4e9</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-06-25
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/cause/blob/8f3b82e734b124b6d4d9d28b9defcce8d3726410/analysis/simulations.Rmd" target="_blank">8f3b82e</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-06-25
</td>
<td>
wflow_publish(files = c(“analysis/about.Rmd”, “analysis/index.Rmd”, “analysis/ldl_cad.Rmd”, “analysis/license.Rmd”,
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="click-here-to-look-at-simulation-results-interactively" class="section level1">
<h1><a href="https://jean-loves-stats.shinyapps.io/simulations/">Click here to look at simulation results interactively</a></h1>
<p>In the document below we describe how to reproduce these using DSC.</p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This page gives instructions for running the simulations that we ran to evaluate CAUSE and will allow you to reproduce the results <a href="https://www.biorxiv.org/content/10.1101/682237v3">the paper</a>. It will hopefully also serve as a resource for explaining those results in more detail and allow you to modify simulations to run your own experiments. There are a lot of parameters that can be changed. We tried to explore a reasonable range of scenarios but there are many more variations that could be tried. Running the simulations involves multiple set-up steps, mostly software installation, and may take a bit of time, especially if you are new to DSC. Additionally, completing the simulations requires several hours of computing and needs to be done on a cluster. If you don’t have access to that kind of computing resource you may still find this section useful and all of our results can be downloaded.</p>
</div>
<div id="set-up" class="section level1">
<h1>Set Up</h1>
<div id="software-installation" class="section level2">
<h2>Software Installation</h2>
<p>Simulations are run using the <code>causeSims</code> R package. The package can be installed with</p>
<pre><code>devtools::install_github(&quot;jean997/causeSims&quot;)</code></pre>
<p>This package relies on a few R packages that are not on CRAN, which you will need to install separately. For detailed installation instructions including installation of supporting R packages for other methods see <a href="https://github.com/jean997/causeSims">here</a>. There you can also find a walk through of a single simulation and explanations of the functions contained in the package.</p>
<p>We conduct simulations using the <a href="https://stephenslab.github.io/dsc-wiki/overview">Dynamical Statisical Comparisons</a> software. You will need to <a href="https://stephenslab.github.io/dsc-wiki/installation.html">have DSC installed</a> and it will be helpful to take a look at <a href="https://stephenslab.github.io/dsc-wiki/first_course/Intro_DSC">the tutorial</a> if you haven’t used DSC before.</p>
</div>
<div id="set-up-the-analysis-directory" class="section level2">
<h2>Set Up The Analysis Directory</h2>
<p>To set up the analysis create a working directory that you will run the simulations in. Make sure all the software in the previous section is installed.</p>
<p>From inside this working directory use the R command</p>
<pre class="r"><code>library(causeSims)
setup_sims()</code></pre>
<p>This function downloads the following files:</p>
<ul>
<li><p>LD data: This is used for simulating summary statistics and is placed into a sub-directory <code>data</code>. The LD data are downloaded from <a href="https://zenodo.org/record/3235780/">here</a> and contain data for HapMap3 SNPs on chromosome 19 estimated from the CEU 1000 Genomes samples.</p></li>
<li><p>DSC files: These contain instructions for DSC on how to simulate data and analyze it. Jump to <a href="#details">this subsection</a> for a detailed description of the DSC files and simulation parameters. If you want to run your own experiments, these files should provide a good starting place for that. The set up function downloads three DSC files. One is for evaluating false positives in the presence of different levels of correlated pleiotropy, one is for evaluating power when there is a true causal effect, and one is for evaluating the robustness of CAUSE to different prior distributions on <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\eta\)</span> (see supplementary note of the paper).</p></li>
<li><p>A configuration file for running DSC on a cluster. This file is set up for the University of Chicago Research Computing Cluster so you will need to change it to match your cluster set up. Even if you are using the U of C cluster, you will need to change the home directory and account lines of this file. Read about using DSC with a compute cluster <a href="https://stephenslab.github.io/dsc-wiki/advanced_course/Remote_Computations">here</a> (We are using the “On Host” option) and modify <code>config.yml</code> as needed.</p></li>
<li><p>Some code for plotting and results summarization</p></li>
<li><p>Data frames of results</p></li>
</ul>
</div>
</div>
<div id="run-dsc" class="section level1">
<h1>Run DSC</h1>
<p>Results in the paper are generated by the following DSC files:</p>
<p>Figure 2a: <code>false_positives.dsc</code> Figure 2b: <code>power.dsc</code> Figure 3: <code>reverse.dsc</code> Supplementary Figure SN4: <code>mixture1.dsc</code></p>
<p>To run DSC use the following commands at the command line</p>
<pre><code>nohup dsc --replicate 100 --host config.yml -c 4 &lt;name&gt;.dsc &gt; &lt;name&gt;.out &amp;</code></pre>
<p>where <code>&lt;name&gt;</code> is replaced by one of the four analysis names above.</p>
<p>Depending on your computing resources and how many jobs you allow to run simultaneously, this could take several hours for each command. Errors can occur that stop DSC, usually a result of interfacing with the cluster. For example, DSC may fail to detect that a job has completed. If this happens, simply restart the failed job using the same command. DSC will pick up where it left off since all results are saved along the way. Running these commands will generate lots of <code>*.err</code> and <code>*.out</code> files all of which can be safely deleted. Results will be in directories named <code>fp</code>, <code>pwr</code>, <code>rev</code> and <code>mix1</code>. You can change these names by changing the <code>output:</code> line of the <code>.dsc</code> files.</p>
</div>
<div id="look-at-results-and-plot-them" class="section level1">
<h1>Look at Results and Plot Them</h1>
<pre class="r"><code>library(causeSims)
library(dplyr)
library(ggplot2)
library(tidyr)
library(purrr)
library(pROC)</code></pre>
<p>Tables of results can be retrieved using the <code>extract_dsc_results</code> function from <code>causeSims</code>. The first argument of this function is a directory name. The function returns a list of data frames. The first gives CAUSE results, the second LCV results, the third results from MR methods run through the <code>MendelianRandomization</code> package. You can toggle extraction of the second two data frames with <code>extract_mr</code> and <code>extract_lcv</code> arguments.</p>
<pre class="r"><code>res_fp &lt;- extract_dsc_results(&quot;fp&quot;)</code></pre>
<p>We can then summarize the results into a plottable format using the <code>summarize_dsc_results</code> function. Here we will use the results that are downloaded with the package.</p>
<pre class="r"><code>res_fp &lt;- readRDS(&quot;sim_results/res_fp.RDS&quot;)
summ_fp &lt;- with(res_fp, summarize_dsc_results(cause, lcv, mr))</code></pre>
<pre><code>Adding missing grouping variables: `cause.qbeta`</code></pre>
<pre class="r"><code>head(summ_fp)</code></pre>
<pre><code># A tibble: 6 x 12
  cause.qbeta     q   tau omega    n1    n2 neffect1 n_sig missing gamma   eta
        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;   &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;
1           2   0.1     0  0.02 12000 12000     1000     0       0     0 0.141
2          10   0.1     0  0.02 12000 12000     1000     0       0     0 0.141
3         100   0.1     0  0.02 12000 12000     1000     0       0     0 0.141
4           2   0.1     0  0.02 12000 40000     1000     0       0     0 0.141
5          10   0.1     0  0.02 12000 40000     1000     0       0     0 0.141
6         100   0.1     0  0.02 12000 40000     1000     0       0     0 0.141
# … with 1 more variable: analysis &lt;chr&gt;</code></pre>
<pre class="r"><code>res_pwr &lt;- readRDS(&quot;sim_results/res_pwr.RDS&quot;)
summ_pwr &lt;- with(res_pwr, summarize_dsc_results(cause, lcv, mr))</code></pre>
<pre><code>Adding missing grouping variables: `cause.qbeta`</code></pre>
<pre class="r"><code>head(summ_pwr)</code></pre>
<pre><code># A tibble: 6 x 12
  cause.qbeta     q   tau omega    n1    n2 neffect1 n_sig missing gamma   eta
        &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;   &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;
1           2     0     0     0 12000 12000     1000     0       0     0     0
2          10     0     0     0 12000 12000     1000     0       0     0     0
3         100     0     0     0 12000 12000     1000     0       0     0     0
4           2     0     0     0 12000 40000     1000     0       0     0     0
5          10     0     0     0 12000 40000     1000     0       0     0     0
6         100     0     0     0 12000 40000     1000     0       0     0     0
# … with 1 more variable: analysis &lt;chr&gt;</code></pre>
<p>This function groups the data by parameter and computes the number of significant results. The <code>p_thresh</code> and <code>gcp_thresh</code> results control the significance threshold. By default these are <span class="math inline">\(p &lt; 0.05\)</span> for MR methods and <span class="math inline">\(\hat{GCP} &gt; 0.6\)</span> for LCV. Now we can plot!</p>
<p>First we plot power with <span class="math inline">\(\tau\)</span> on the horizontal axis and the number of times each method rejected the null hypothesis on the vertical axis.</p>
<pre class="r"><code>wong_colors &lt;- c(&quot;#000000&quot;, &quot;#E69F00&quot;, &quot;#56B4E9&quot;, &quot;#009E73&quot;, &quot;#0072B2&quot;, 
                &quot;#D55E00&quot;, &quot;#CC79A7&quot;, &quot;#F0E442&quot;)
methods =c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
            &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;)
p_thresh = 0.05
gcp_thresh = 0.6
shp &lt;- c(16, 3, 4, 15, 17, 25, 23, 8)[1:length(methods)]
ix &lt;- match(methods,c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                          &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;, &quot;lcv_mean&quot;) )
cols &lt;- wong_colors[ix]</code></pre>
<pre class="r"><code>#plot
pwr &lt;- summ_pwr %&gt;%
    filter(q == 0) %&gt;%
    filter(analysis %in% methods) %&gt;%
    mutate(n1 = as.character(n1),
           n1 = recode(n1, &quot;12000&quot; = &quot;M low power&quot;, 
                       &quot;40000&quot; = &quot;M high power&quot;), 
           n2 = as.character(n2),
           n2 = recode(n2, &quot;12000&quot; = &quot;Y low power&quot;, 
                       &quot;40000&quot; = &quot;Y high power&quot;),
           analysis = factor(analysis, levels = methods),
           analysis = recode(analysis, cause_10 = &quot;CAUSE&quot;,
                             egger_MR = &quot;Egger Regression&quot;, 
                             ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                             mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr = &quot;GSMR&quot;)) %&gt;%
    ungroup() %&gt;%
    ggplot(.) +
    geom_line(aes(x=tau, y=n_sig/100, color=analysis)) +
    geom_point(aes(x=tau, y=n_sig/100, color=analysis, shape=analysis), size=3) +
    scale_color_manual(values=cols) +
    scale_shape_manual(values= shp) +
    scale_y_continuous(limits=c(0, 1)) + 
    geom_hline(yintercept=p_thresh, linetype=2) +
    xlab(expression(&quot;Percentage of Y heritability explained by M (&quot; ~ tau ~ &quot;=&quot; ~ gamma^2 ~ h[M]^2/h[Y]^2 ~ &quot;)&quot;)) +
    ylab(&quot;Percent Significant&quot;) +
    #theme_bw() +
    ggtitle(&quot;Power&quot;) +
    theme_bw() + 
    theme(legend.title = element_blank(),
          strip.text = element_text(size=14),
          axis.title = element_text(size=16),
          legend.text = element_text(size=12),
          plot.title = element_text(size=20),
          legend.position = &quot;bottom&quot;) +
    guides(color=guide_legend(nrow=1),
           linetype=guide_legend(nrow=1),
           shape=guide_legend(nrow=1) )  +
    facet_wrap(~n1*n2, nrow=1)
pwr</code></pre>
<p><img src="figure/simulations.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>CAUSE has lower power than the other methods, except the weighted mode and Egger regression. When the power for trait <span class="math inline">\(M\)</span> is low and the power for trait <span class="math inline">\(Y\)</span> is high, CAUSE has better power for large causal effects. This is expected since CAUSE uses a more stringent criterion to conclude that the data are consistent with a causal effect. It is able to achieve better power when <span class="math inline">\(M\)</span> is under-powered and <span class="math inline">\(Y\)</span> is well powered because it uses all variants genome-wide rather than only those that are highly significant for trait <span class="math inline">\(M\)</span>. This is an important setting because it may be applicable to scenarios when <span class="math inline">\(M\)</span> is a biomarker or molecular trait and <span class="math inline">\(Y\)</span> is a complex triat.</p>
<div id="false-positives" class="section level3">
<h3>False Positives</h3>
<p>Next we plot false positives. Our simulations include two values of <span class="math inline">\(\eta\)</span> (<span class="math inline">\(\sqrt{0.05}\)</span> and <span class="math inline">\(\sqrt{0.02}\)</span>). Although we only had space to present one set of results in the paper, we will make plots for both here. We plot the proportion of <span class="math inline">\(M\)</span> effect variants that display correlated pleiotropy, <span class="math inline">\(q\)</span>, on the horizontal axis and the number of discoveries on the vertical axis. The bottom row of plots are for the larger value of <span class="math inline">\(\eta\)</span> and are the ones shown in the paper. For convenience, in extracting results, we copy the results for <span class="math inline">\(q=0\)</span>.</p>
<pre class="r"><code>z1 &lt;-  filter(summ_fp, omega == 0 &amp; tau == 0 &amp; q == 0) %&gt;% mutate(omega = 0.05)
z2 &lt;-  filter(summ_fp, omega == 0 &amp; tau == 0 &amp; q == 0) %&gt;% mutate(omega = 0.02)
summ_fp_copy &lt;- bind_rows(summ_fp, z1, z2) %&gt;%
        filter(!(q == 0 &amp; omega == 0 &amp; tau == 0))</code></pre>
<pre class="r"><code>fp &lt;- summ_fp_copy %&gt;%
    filter(tau == 0) %&gt;% # &amp; omega == input$roc_effect) %&gt;%
    filter(analysis %in% methods) %&gt;%
    mutate(n1 = as.character(n1),
           n1 = recode(n1, &quot;12000&quot; = &quot;M low power&quot;, 
                       &quot;40000&quot; = &quot;M high power&quot;), 
           n2 = as.character(n2),
           n2 = recode(n2, &quot;12000&quot; = &quot;Y low power&quot;, 
                       &quot;40000&quot; = &quot;Y high power&quot;),
           omega = recode(omega, `0.02` = &quot;Weak shared factor effect&quot;, `0.05` = &quot;Strong shared factor effect&quot;),
           omega = factor(omega, levels = c(&quot;Weak shared factor effect&quot;, &quot;Strong shared factor effect&quot;)),
           analysis = factor(analysis, levels = methods),
           analysis = recode(analysis, cause_10 = &quot;CAUSE&quot;, causeeqg_10 = &quot;CAUSE-eqg&quot;,
                             cause_fixedest_10 = &quot;CAUSE-fe&quot;, cause_fixedoracle_10 = &quot;CAUSE-fo&quot;,
                             egger_MR = &quot;Egger Regression&quot;, 
                             ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                             mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr = &quot;GSMR&quot;)) %&gt;%
    ggplot(.) + 
      geom_line(aes(x=q, y=n_sig/100, color=analysis)) + 
      geom_point(aes(x=q, y=n_sig/100, color=analysis, shape=analysis), size=3) + 
      geom_hline(yintercept=as.numeric(p_thresh), linetype=2) + 
      scale_color_manual(values=cols) +
      scale_shape_manual(values= shp) +
      scale_y_continuous(limits=c(0, 1)) + 
      geom_hline(yintercept=0.05, linetype=2) + 
      xlab(&quot;Proportion of variants acting through U (q)&quot;) + 
      ylab(&quot;Percent Significant&quot;) + 
      ggtitle(&quot;False Positive Rate&quot;) +
      theme_bw() + 
      theme(legend.title = element_blank(), 
          strip.text = element_text(size=14), 
          axis.title = element_text(size=16), 
          legend.text = element_text(size=12), 
          plot.title = element_text(size=20),
          legend.position = &quot;bottom&quot;) + 
      guides(color=guide_legend(nrow=1), 
           linetype=guide_legend(nrow=1),
           shape=guide_legend(nrow=1) )  + 
      facet_wrap(~omega*n1*n2, nrow=2)
fp</code></pre>
<p><img src="figure/simulations.Rmd/plot_fp-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_fp-1">
Past versions of plot_fp-1.png
</button>
</p>
<div id="fig-plot_fp-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/cause/blob/b1154146bef8ee5f0a4b1471490734da51ba2f29/docs/figure/simulations.Rmd/plot_fp-1.png" target="_blank">b115414</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-07-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>In all cases CAUSE has a lower false positive rate than alternative methods except the weighted mode and Egger regression. The false positive rate increases for large values of <span class="math inline">\(q\)</span> and is worse when the effect of the shared factor is stronger.</p>
</div>
<div id="roc-curves" class="section level3">
<h3>ROC Curves</h3>
<p>We use ROC curves to evaluate how well each method is able to distinguish simulations with a causal effect of size <span class="math inline">\(\gamma = \sqrt{0.05}\)</span> from simulations with a shared factor with an equal effect acconting for 30% of the trait <span class="math inline">\(M\)</span> vairants. Both of these parameters can be modified. Processing hte data to create the ROC curves is a bit complicated. We also add in LCV for these compairsons.</p>
<pre class="r"><code>q = 0.3 #Can modify to 0, 0.1, 0.2, 0.4 or 0.5 to see alternative results

roc_effect = 0.05 # Also try 0.02

my_q &lt;-q
if(my_q == 0){
    eff2 &lt;- 0
}else{
    eff2 &lt;- roc_effect
}

methods =c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                                &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;, &quot;lcv_mean&quot;)
shp &lt;- c(16, 3, 4, 15, 17, 25, 23, 8)[1:length(methods)]
ix &lt;- match(methods,c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                          &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;, &quot;lcv_mean&quot;) )
cols &lt;- wong_colors[ix]
  

res_l &lt;- list(res_fp, res_pwr)
cdf &lt;- purrr::map_df(res_l, function(r){r$cause})
ldf &lt;- purrr::map_df(res_l, function(r){r$lcv})
mdf &lt;- purrr::map_df(res_l, function(r){r$mr})

#Combined data 
dat &lt;- roc_data(cdf, ldf, mdf)
head(dat)</code></pre>
<pre><code>  simulate.output.file   q omega tau    n1    n2  analysis         stat
1  simulate/simulate_1 0.1  0.02   0 12000 12000  cause_10 7.459915e-01
2  simulate/simulate_1 0.1  0.02   0 12000 12000 cause_100 7.730339e-01
3  simulate/simulate_1 0.1  0.02   0 12000 12000   cause_2 6.097030e-01
4  simulate/simulate_2 0.2  0.02   0 12000 12000  cause_10 6.529461e-13
5  simulate/simulate_2 0.2  0.02   0 12000 12000 cause_100 0.000000e+00
6  simulate/simulate_2 0.2  0.02   0 12000 12000   cause_2 0.000000e+00</code></pre>
<pre class="r"><code>#Combined summary for adding dots to graph
res &lt;- summarize_dsc_results(cdf, ldf, mdf, 
                             p_thresh=p_thresh, gcp_mean_thresh=gcp_thresh)</code></pre>
<pre><code>Adding missing grouping variables: `cause.qbeta`</code></pre>
<pre class="r"><code>pw &lt;- filter(res, tau == roc_effect ) %&gt;% ungroup()
  f &lt;- filter(res, omega == eff2 &amp; q == my_q &amp; tau ==0) %&gt;% ungroup()
  df &lt;- select(pw, analysis, n_sig, n1, n2) %&gt;% 
    rename(power = n_sig) %&gt;% 
    full_join(., f, c(&quot;analysis&quot;, &quot;n1&quot;, &quot;n2&quot;)) %&gt;% 
    rename(fp = n_sig) %&gt;% 
    select(analysis, power, fp, n1, n2) %&gt;%
    filter(analysis %in% methods) %&gt;%
    rename(method=analysis) %&gt;%
    mutate(n1 = as.character(n1),
           n1 = recode(n1, &quot;12000&quot; = &quot;M low power&quot;, 
                       &quot;40000&quot; = &quot;M high power&quot;), 
           n2 = as.character(n2),
           n2 = recode(n2, &quot;12000&quot; = &quot;Y low power&quot;, 
                       &quot;40000&quot; = &quot;Y high power&quot;),
           method = factor(method, levels = methods),
           method = recode(method, cause_10 = &quot;CAUSE&quot;, egger_MR = &quot;Egger Regression&quot;, 
                           ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                           mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr = &quot;GSMR&quot;,
                           lcv_mean = &quot;LCV GCP&quot;)) 
  
  
  
n1s &lt;- c(12000, 12000, 40000, 40000)
n2s &lt;- c(12000, 40000, 12000, 40000)

#Data frame of points along ROC curves  
roc_df &lt;-map_dfr(seq_along(n1s), function(i){ 
    my_n1 &lt;- n1s[i] 
    my_n2 &lt;- n2s[i]
    rocs &lt;- map_dfr(methods, function(m){
      cat(m, &quot; &quot;)
      xx &lt;- filter(dat, n1 == my_n1 &amp; n2 == my_n2 &amp; analysis == m &amp; 
                     ( tau == roc_effect | (omega == eff2 &amp; q == my_q &amp; tau == 0))) 
      xx$stat[is.na(xx$stat)] &lt;- 0
      with(xx, roc(predictor = stat, response = tau !=0, 
                   levels=c(FALSE, TRUE), direction=&quot;&lt;&quot;)) %&gt;% 
        with(., data.frame(thresh = thresholds, spec = specificities, 
                           sens = sensitivities, method = m, 
                           n1 = my_n1, n2 = my_n2, auc=rep(auc, length(thresholds)),
                           stringsAsFactors=F)) %&gt;% return()
      
    })
  })%&gt;%   
  mutate(n1 = recode(n1, `12000` = &quot;M low power&quot;, 
                       `40000` = &quot;M high power&quot;), 
           n2 = recode(n2, `12000` = &quot;Y low power&quot;, 
                       `40000` = &quot;Y high power&quot;), 
           n1 = factor(n1, levels=c(&quot;M low power&quot;, &quot;M high power&quot;)),
           n2 = factor(n2, levels=c(&quot;Y low power&quot;, &quot;Y high power&quot;)))</code></pre>
<pre><code>cause_10  ivw_MR  egger_MR  gsmr  wm_MR  mrp  mbe_MR_1  lcv_mean  cause_10  ivw_MR  egger_MR  gsmr  wm_MR  mrp  mbe_MR_1  lcv_mean  cause_10  ivw_MR  egger_MR  gsmr  wm_MR  mrp  mbe_MR_1  lcv_mean  cause_10  ivw_MR  egger_MR  gsmr  wm_MR  mrp  mbe_MR_1  lcv_mean  </code></pre>
<pre class="r"><code>#Plot!  
plt &lt;- roc_df %&gt;% arrange(sens) %&gt;%
          mutate(
           method = factor(method, levels = methods),
           method = recode(method, cause_10 = &quot;CAUSE&quot;, egger_MR = &quot;Egger Regression&quot;, 
                           ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                           mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr = &quot;GSMR&quot;,
                           lcv_mean = &quot;LCV GCP&quot;))  %&gt;%
    ggplot(.) + 
    geom_line(aes(x=1-spec, y=sens, col=method, group=method)) + 
    geom_point(aes(x=fp/100, y = power/100, col=method, shape=method, fill = method), size =3 , data=df) + 
    scale_color_manual(values=cols) + 
    scale_fill_manual(values=cols) + 
    scale_shape_manual(values = shp) + 
    xlab(&quot;1-Specificity (False Positive Rate)&quot;) + ylab(&quot;Sensitivity (Power)&quot;) + 
    ggtitle(&quot;False Positive Rate vs Power&quot;) +
    theme_bw() + 
    theme(legend.title = element_blank(), 
          strip.text = element_text(size=14), 
          axis.title = element_text(size=16), 
          plot.title = element_text(size=20),
          legend.text = element_text(size=12), 
          legend.position = &quot;bottom&quot;) + 
    guides(color=guide_legend(nrow=1), 
           linetype=guide_legend(nrow=1),
           shape=guide_legend(nrow=1) 
    )  + 
    facet_wrap(~n1*n2, ncol=4)
plt</code></pre>
<p><img src="figure/simulations.Rmd/roc-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>  auc_tab &lt;- filter(roc_df, thresh==-Inf) %&gt;%
      select(-thresh, -spec, -sens) %&gt;%
      mutate(q=q)
  auc_tab &lt;- auc_tab %&gt;%
    mutate(method = recode(method, cause_10 = &quot;CAUSE&quot;, egger_MR = &quot;Egger Regression&quot;, 
                           ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                           mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr = &quot;GSMR&quot;,
                           lcv_mean = &quot;LCV GCP&quot;),
      class = case_when(n1==&quot;M low power&quot; &amp; n2==&quot;Y low power&quot; ~ &quot;LL&quot;, 
                             n1==&quot;M low power&quot; &amp; n2==&quot;Y high power&quot; ~ &quot;LH&quot;, 
                             n1==&quot;M high power&quot; &amp; n2==&quot;Y low power&quot; ~ &quot;HL&quot;, 
                             n1==&quot;M high power&quot; &amp; n2==&quot;Y high power&quot; ~ &quot;HH&quot;
                             )) %&gt;%
    select(-n1, -n2, -q) %&gt;%
    spread(class, auc) %&gt;%
    mutate(q = q)
 
auc_tab</code></pre>
<pre><code>            method     HH     HL     LH     LL   q
1            CAUSE 1.0000 0.9970 0.9998 0.9925 0.3
2 Egger Regression 0.9653 0.8094 0.5674 0.5470 0.3
3             GSMR 1.0000 1.0000 0.8628 0.7849 0.3
4   IVW Regression 1.0000 1.0000 0.9579 0.8636 0.3
5          LCV GCP 0.9795 0.9862 0.9777 0.9601 0.3
6        MR-PRESSO 1.0000 1.0000 0.9216 0.8240 0.3
7  Weighted Median 1.0000 0.9993 0.9174 0.8121 0.3
8    Weighted Mode 0.9976 0.9259 0.8731 0.7167 0.3</code></pre>
</div>
<div id="reverse-direction-effects" class="section level3">
<h3>Reverse direction effects</h3>
<p>These results show the performance of each method when the data are simulated from the model <span class="math inline">\(Y \rightarrow M\)</span> but we test for an effect of <span class="math inline">\(M\)</span> on <span class="math inline">\(Y\)</span>.</p>
<pre class="r"><code>res_rev &lt;- readRDS(&quot;sim_results/res_rev.RDS&quot;)
summ_rev &lt;- with(res_rev, summarize_dsc_results(cause, lcv, mr))</code></pre>
<pre><code>Adding missing grouping variables: `cause.qbeta`</code></pre>
<pre class="r"><code>methods =c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
            &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;)
shp &lt;- c(16, 3, 4, 15, 17, 25, 23, 8)[1:length(methods)]
ix &lt;- match(methods,c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                          &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;, &quot;lcv_mean&quot;) )
cols &lt;- wong_colors[ix]

rev &lt;- summ_rev %&gt;%
    filter(analysis %in% methods) %&gt;%
    mutate(analysis = factor(analysis, levels = c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                                                  &quot;mbe_MR_1&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;)),
           analysis = recode(analysis, cause_10 = &quot;CAUSE&quot;, egger_MR = &quot;Egger Regression&quot;, 
                             ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                             mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;)) %&gt;%
    ungroup() %&gt;%
    ggplot(.) +
    geom_line(aes(x=neffect1, y=n_sig/100, color=analysis)) +
    geom_point(aes(x=neffect1, y=n_sig/100, color=analysis, shape=analysis), size=3) +
    scale_color_manual(values=cols) +
    scale_shape_manual(values= shp) +
    scale_y_continuous(limits=c(0, 1)) + 
    scale_x_continuous(breaks=c(250, 500, 1000)) + 
    geom_hline(yintercept=as.numeric(p_thresh), linetype=2) +
    xlab(expression(&quot;Number of M effect variants independent of Y&quot;)) +
    ylab(&quot;Percent Significant&quot;) +
    theme_bw() +
    ggtitle(&quot;Robustness to Reverse Effects&quot;) +
    theme(legend.title = element_blank(),
          strip.text = element_text(size=14),
          axis.title = element_text(size=16),
          legend.text = element_text(size=12),
          legend.position = &quot;bottom&quot;, 
          plot.title = element_text(size=20)) +
    guides(color=guide_legend(nrow=2),
           linetype=guide_legend(nrow=2),
           shape=guide_legend(nrow=2) ) 
 
rev</code></pre>
<p><img src="figure/simulations.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="causal-effect-and-shared-factor" class="section level3">
<h3>Causal effect and shared factor</h3>
<p>These are the results in Section SN6.3</p>
<pre class="r"><code>methods =c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
            &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;)
shp &lt;- c(16, 3, 4, 15, 17, 25, 23, 8)[1:length(methods)]
ix &lt;- match(methods,c(&quot;cause_10&quot;, &quot;ivw_MR&quot;, &quot;egger_MR&quot;, 
                          &quot;gsmr&quot;, &quot;wm_MR&quot;, &quot;mrp&quot;, &quot;mbe_MR_1&quot;, &quot;lcv_mean&quot;) )
cols &lt;- wong_colors[ix]

res_mix &lt;- readRDS(&quot;sim_results/res_mix1.RDS&quot;)
summ_mix &lt;- with(res_mix, summarize_dsc_results(cause, lcv, mr, direction = &quot;pos&quot;))</code></pre>
<pre><code>Adding missing grouping variables: `cause.qbeta`</code></pre>
<pre class="r"><code>plt_mix &lt;- summ_mix %&gt;%
    filter(analysis %in% methods) %&gt;%
    mutate(analysis = factor(analysis, levels = methods),
           analysis = recode(analysis, cause_10 = &quot;CAUSE&quot;, egger_MR = &quot;Egger Regression&quot;, 
                             ivw_MR = &quot;IVW Regression&quot;, mbe_MR_1 = &quot;Weighted Mode&quot;, 
                             mrp = &quot;MR-PRESSO&quot;, wm_MR = &quot;Weighted Median&quot;, gsmr=&quot;GSMR&quot;), 
           n1 = as.character(n1),
           n1 = recode(n1, &quot;12000&quot; = &quot;M low power&quot;, 
                       &quot;40000&quot; = &quot;M high power&quot;), 
           n2 = as.character(n2),
           n2 = recode(n2, &quot;12000&quot; = &quot;Y low power&quot;, 
                       &quot;40000&quot; = &quot;Y high power&quot;),
           q = recode(q, `0.1` = &quot;q = 0.1&quot;, `0.3` = &quot;q = 0.3&quot;)) %&gt;%
    ungroup() %&gt;%
    ggplot(.) +
    geom_line(aes(x=eta, y=n_sig/100, color=analysis)) +
    geom_point(aes(x=eta, y=n_sig/100, group = analysis, color=analysis, shape=analysis)) +
    scale_color_manual(values=cols) +
    scale_shape_manual(values= shp) +
    xlab(expression(eta)) +
    ylab(&quot;Percent Significant&quot;) +
    theme_bw() + 
    theme(legend.title = element_blank(),
          strip.text = element_text(size=14),
          axis.title = element_text(size=16),
          legend.text = element_text(size=12),
          plot.title = element_text(size=16),
          legend.position = &quot;bottom&quot;) +
    guides(color=guide_legend(nrow=1),
           linetype=guide_legend(nrow=1),
           shape=guide_legend(nrow=1) )  +
    facet_wrap(~q*n1*n2, nrow=2) 
plt_mix</code></pre>
<p><img src="figure/simulations.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="whats-in-the-dsc-files" class="section level1">
<h1>What’s in the DSC Files?</h1>
<p>We will now go in detail through one of the DSC files so you can see what is happening and how to make changes to run your own experiments. The DSC files are all identical except for the first few module. <code>reverse.dsc</code> has an extra module that reverses the data (switching traits <span class="math inline">\(M\)</span> and <span class="math inline">\(Y\)</span>).</p>
<p>The DSC statement is at the top:</p>
<pre><code>DSC:
    define:
      mr: ivw_MR, egger_MR, wm_MR, mbe_MR, mrp, gsmr
    run: simulate*data_summ,
         simulate*LCV,
         simulate*mr,
         simulate*cause_params*cause
    replicate: 100
    output: fp</code></pre>
<p>This statement tells DSC what to run. In <code>define</code> we define a group of modules. These will turn up later in the file. In <code>run</code> we tell DSC what to run. For example, <code>simulate*gw_sig</code> means that DSC should run the <code>simulate</code> module followed by the <code>gw_sig</code> module which will use the results of the <code>simulate</code> module. <code>replicate</code> indicates the number of replicates to run (though this is over-ridden by the <code>--replicate</code> flag in the DSC call above) and <code>output</code> names the output directory.</p>
<p>The rest of the document defines modules. All of our modules use R code and most of them rely on the <code>causeSims</code> R package.</p>
<div id="simulation-modules" class="section level2">
<h2>Simulation Modules</h2>
<p>The simulation module:</p>
<pre><code>simulate: R(library(causeSims);
            snps &lt;- readRDS(&quot;data/chr19_snpdata_hm3only.RDS&quot;);
            evd_list &lt;- readRDS(&quot;data/evd_list_chr19_hm3.RDS&quot;);
            dat &lt;- sum_stats(snps, evd_list,
                  n_copies = 30,
                  n1 = n1, n2=n2,
                  h1=h1, h2=h1,
                  neffect1 = neffect1, neffect2 = neffect2,
                  tau = tau, omega = omega, q = q,
                  cores = 4, ld_prune_pval_thresh = 0.01,
                  r2_thresh = 0.1))

     q: 0.1, 0.2, 0.3, 0.4, 0.5
     omega: 0.02, 0.05
     tau: 0
     n1: 12000, 40000
     n2: 12000, 40000
     neffect1: 1000
     neffect2: 1000
     h1: 0.25
     h2: 0.25
     $sim_params: c(q = q, omega = omega, tau = tau,  h1 = h1, h2 = h2, n1 = n1, n2 = n2, neffect1 = neffect1, neffect2 =neffect2)
     $dat: dat</code></pre>
<p>The R code contained in <code>R()</code> is the guts of this module. A data object named <code>dat</code> is generated using the <code>sum_stats</code> function from the <code>causeSims</code> R pacakge. Below this, we specify parameters used to simulate the data. Most important are <code>q</code>, <code>omega</code>, and <code>tau</code>. <code>q</code> is the proportion of variants acting through <span class="math inline">\(U\)</span> while <code>omega</code> and <code>tau</code> are standardized versions of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\eta\)</span> in the paper. <code>tau</code> is the proportion of <span class="math inline">\(Y\)</span> heritability explained by the causal effect of <span class="math inline">\(M\)</span> on <span class="math inline">\(Y\)</span> times the sign of <span class="math inline">\(\gamma\)</span>, <span class="math display">\[
\tau = sign(\gamma)\frac{\gamma^2 h^2_{M}}{h^{2}_{Y}},
\]</span> and <code>omega</code> is the proportion of <span class="math inline">\(Y\)</span> heritability explained by <span class="math inline">\(U\)</span>, divided by <span class="math inline">\(q\)</span> and multiplied by the sign of <span class="math inline">\(\eta\)</span>. <span class="math display">\[
\omega = sign(\eta)\frac{\eta^2 h^2_{M}}{h^{2}_{Y}}
\]</span> This parameterization is sometimes more convenient for running simulations that using <span class="math inline">\(\eta\)</span> and <span class="math inline">\(\gamma\)</span>. However, the <code>sum_stats</code> function will accept either <code>omega</code> and <code>tau</code> arguments or <code>eta</code> and <code>gamma</code>. Note that here we have specified that the function can use 4 cores. This should match the <code>n_cpu</code> parameter in the <code>simulate</code> section of the config file. The <code>sum_stats</code> function generates data both with and without LD. In this file we only analyze the “with LD” data but we will describe how to make modifications to analyze both or only the “no LD” data.</p>
<p>The <code>data_summ</code> module is a helper module. This module computes the GCP (a parameter defined by LCV (O’Connor and Price 2019)) and computes the number of genome-wide significant variants. It also stores all the simulation parameters which makes it easier to extract them. Extracting parameters from the <code>simulate</code> module will require reading in the entire data set, objects created by the <code>data_summ</code> module are smaller and faster to read.</p>
<pre><code>data_summ: R(library(causeSims);
           m_sig_nold &lt;- with($(dat), sum(p_value_nold &lt; thresh ));
           y_sig_nold &lt;- with($(dat), sum(2*pnorm(-abs(beta_hat_2_nold/seb2)) &lt; thresh ));
           m_sig &lt;- with($(dat), sum(p_value &lt; thresh &amp; ld_prune==TRUE));

          params &lt;- $(sim_params);
          tau &lt;- as.numeric(params[&quot;tau&quot;]);
          omega &lt;- as.numeric(params[&quot;omega&quot;]);
          q &lt;- as.numeric(params[&quot;q&quot;]);
          h1 &lt;- as.numeric(params[&quot;h1&quot;]);
          h2 &lt;- as.numeric(params[&quot;h2&quot;]);
          neffect1 &lt;- as.numeric(params[&quot;neffect1&quot;]);
          neffect2 &lt;- as.numeric(params[&quot;neffect2&quot;]);
          gamma &lt;- sqrt(tau*sum(h2)/sum(h1))*sign(tau);
          eta &lt;- sqrt(abs(omega)*sum(h2)/sum(h1))*sign(omega);

          #LCV parameters
          if(q == 0 &amp; tau == 0){
            q.1 &lt;- q.2 &lt;- 0;
            gcp &lt;- 0;
          }else if(q == 0){
            q.1 &lt;- 1;
            q.2 &lt;- 0;
            gcp &lt;- 1;
          }else{
            q.1 &lt;- sqrt(q);
            q.2 &lt;- eta * sqrt(q) * sqrt(h1) / sqrt(h2);
            gcp = (log(q.2^2) - log(q.1^2)) / (log(q.2^2) + log(q.1^2));
          };
          )
    thresh: 5e-8
    $m_sig: m_sig
    $m_sig_nold: m_sig_nold
    $y_sig_nold: y_sig_nold
    $q: q
    $omega: omega
    $tau: tau
    $gamma: gamma
    $eta: eta
    $h1: h1
    $h2: h2
    $n1: as.numeric(params[&quot;n1&quot;])
    $n2: as.numeric(params[&quot;n2&quot;])
    $neffect1: as.numeric(params[&quot;neffect1&quot;])
    $neffect2: as.numeric(params[&quot;neffect2&quot;])
    $lcv_q1: q.1
    $lcv_q2: q.2
    $lcv_gcp: gcp</code></pre>
</div>
<div id="cause-modules" class="section level2">
<h2>CAUSE Modules</h2>
<p>We have divided CAUSE analysis into parameter estimation and model fitting steps. This makes it easier to experiment with changes only to the parameter estimation step.</p>
<pre><code>cause_params: R(library(causeSims);
                p &lt;- cause_params_sims($(dat), null_wt = null_wt, no_ld=FALSE);
                )
    null_wt: 10
    $cause_params_ld: p

cause: R(library(causeSims);
         library(cause);
         cause_res &lt;- cause_sims($(dat), $(cause_params_ld), no_ld = FALSE, 
                                qalpha = 1, qbeta = qbeta);
         z &lt;- -1*summary(cause_res)$z;
         p &lt;- pnorm(-z);
         qs &lt;- summary(cause_res)$quants)
    qbeta: 10, 100, 2
    $cause_res: cause_res
    $sigma_g: cause_res$sigma_g
    $z: z
    $p: p
    $eta_2_upper: qs[[1]][3,2]
    $q_2_upper: qs[[1]][3,3]
    $eta_2_lower: qs[[1]][2,2]
    $q_2_lower: qs[[1]][2,3]
    $eta_2_med: qs[[1]][1,2]
    $q_2_med: qs[[1]][1,3]
    $eta_3_upper: qs[[2]][3,2]
    $q_3_upper: qs[[2]][3,3]
    $gamma_3_upper: qs[[2]][3,1]
    $eta_3_lower: qs[[2]][2,2]
    $q_3_lower: qs[[2]][2,3]
    $gamma_3_lower: qs[[2]][2,1]
    $eta_3_med: qs[[2]][1,2]
    $q_3_med: qs[[2]][1,3]
    $gamma_3_med: qs[[2]][1,1]</code></pre>
<p>Both modules use wrapper functions from the <code>causeSims</code> package. These are very simple wrappers that make it easy to run CAUSE on the simulated data format. Note that here we use <code>no_ld = FALSE</code> in both modules. If you want to use the data that is simulated without LD instead you can change <code>no_ld = TRUE</code>.</p>
</div>
<div id="alternative-mr-modules" class="section level2">
<h2>Alternative MR Modules</h2>
<p>There are several modules for alternative MR methods. Modules for IVW regression, Egger regression, MR-PRESSO, and the weighted median method are very similar and rely on wrappers from <code>causeSims</code>. We can just look at one of these.</p>
<pre><code>ivw_MR: R(library(causeSims);
       res &lt;- ivw_re_nonome_MR($(dat), p_val_thresh=thresh, no_ld = no_ld);
       )
    thresh: 5e-8
    no_ld: FALSE
    $est: res$est
    $z: res$z
    $p: res$p</code></pre>
<p>To run on the data without LD change the <code>no_ld</code> parameter to <code>no_ld: TRUE</code> or <code>no_ld: TRUE, FALSE</code> to run both.</p>
<p>The GSMR module is a little more complicated</p>
<pre><code>gsmr: R(library(causeSims);
        evd_list &lt;- readRDS(&quot;data/evd_list_chr19_hm3.RDS&quot;);
        res &lt;- gsmr_sims($(dat), evd_list, p_val_thresh  = 5e-8, no_ld = FALSE);
        if(!is.null(res)){
           z &lt;- res$bxy/res$bxy_se;
           est &lt;- res$bxy;
           p &lt;- res$bxy_pval;
        }else{
           z &lt;- est &lt;- p &lt;-  NA;
        }
      )
    thresh: 5e-8
    $z: z
    $p: p
    $est: est
    $gsmr: res</code></pre>
<p>GSMR requires the LD structure and sometimes returns errors. In this case the <code>gsmr_sims</code> wrapper returns NULL and we set results to missing. If <code>no_ld = TRUE</code> the LD data in <code>evd_list</code> is simply ignored.</p>
</div>
<div id="lcv-module" class="section level2">
<h2>LCV Module</h2>
<p>Results from LCV are not compared with CAUSE in the paper because LCV is performing a different test than other MR methods. However, we have implemented wrappers and a module to run LCV.</p>
<pre><code>LCV: R(library(causeSims);
       res &lt;- try(lcv_sims($(dat),no_ld = no_ld, sig.thresh = thresh));
       if(class(res) == &quot;try-error&quot;){
            res &lt;- list(pval.gcpzero.2tailed=NA, gcp.pm = NA, gcp.pse = NA);
        };
       )
    thresh: 30
    no_ld: FALSE
    $p: res$pval.gcpzero.2tailed
    $gcp_mean: res$gcp.pm
    $gcp_pse: res$gcp.pse
    $gcp_obj: res</code></pre>
<p>When LCV returns errors, we set results to missing.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] pROC_1.16.2     purrr_0.3.4     tidyr_1.1.0     ggplot2_3.3.2  
[5] dplyr_1.0.0     causeSims_0.1.0 workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] httr_1.4.1                   jsonlite_1.7.0              
 [3] viridisLite_0.3.0            RcppParallel_5.0.2          
 [5] assertthat_0.2.1             mixsqp_0.3-43               
 [7] arrangements_1.1.8           yaml_2.2.1                  
 [9] robustbase_0.93-6            globals_0.12.5              
[11] numDeriv_2016.8-1.1          pillar_1.4.4                
[13] backports_1.1.8              lattice_0.20-41             
[15] glue_1.4.1                   digest_0.6.25               
[17] promises_1.1.1               iterpc_0.4.2                
[19] colorspace_1.4-1             MRPRESSO_1.0                
[21] plyr_1.8.6                   htmltools_0.5.0             
[23] httpuv_1.5.4                 Matrix_1.2-18               
[25] pkgconfig_2.0.3              invgamma_1.1                
[27] listenv_0.8.0                scales_1.1.1                
[29] intervals_0.15.2             cause_1.0.0.0267            
[31] whisker_0.4                  later_1.1.0.1               
[33] git2r_0.27.1                 tibble_3.0.2                
[35] MendelianRandomization_0.4.2 gmp_0.6-0                   
[37] farver_2.0.3                 generics_0.0.2              
[39] ellipsis_0.3.1               withr_2.2.0                 
[41] furrr_0.1.0                  ashr_2.2-47                 
[43] lazyeval_0.2.2               cli_2.0.2                   
[45] magrittr_1.5                 crayon_1.3.4                
[47] evaluate_0.14                fansi_0.4.1                 
[49] fs_1.4.2                     future_1.18.0               
[51] truncnorm_1.0-8              RcppZiggurat_0.1.5          
[53] tools_4.0.2                  loo_2.3.0                   
[55] data.table_1.12.8            hms_0.5.3                   
[57] lifecycle_0.2.0              matrixStats_0.56.0          
[59] stringr_1.4.0                plotly_4.9.2.1              
[61] munsell_0.5.0                irlba_2.3.3                 
[63] compiler_4.0.2               rlang_0.4.6                 
[65] grid_4.0.2                   rstudioapi_0.11             
[67] rjson_0.2.20                 htmlwidgets_1.5.1           
[69] labeling_0.3                 rmarkdown_2.3               
[71] gtable_0.3.0                 codetools_0.2-16            
[73] R6_2.4.1                     gridExtra_2.3               
[75] knitr_1.29                   utf8_1.1.4                  
[77] rprojroot_1.3-2              readr_1.3.1                 
[79] stringi_1.4.6                parallel_4.0.2              
[81] SQUAREM_2020.3               Rcpp_1.0.5                  
[83] vctrs_0.3.1                  DEoptimR_1.0-8              
[85] tidyselect_1.1.0             xfun_0.15                   </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
